src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/CSharpConvertToRecordRefactoringProvider.cs

```diff
@@ -0,0 +1,623 @@
+
+            // remove converted properties and reformat other methods
+            var membersToKeep = GetModifiedMembersForPositionalRecord(originalType, originalDeclarationNode, semanticModel, propertiesToMove, cancellationToken);
+
+            var changedTypeDeclaration = SyntaxFactory.RecordDeclaration(
+                originalType.TypeKind == TypeKind.Class
+                    ? SyntaxKind.RecordDeclaration
+                    : SyntaxKind.RecordStructDeclaration,
+                originalDeclarationNode.GetAttributeLists(),
+                originalDeclarationNode.GetModifiers(),
+                SyntaxFactory.Token(SyntaxKind.RecordKeyword),
+                originalType.TypeKind == TypeKind.Class
+                    ? default
+                    : SyntaxFactory.Token(SyntaxKind.StructKeyword),
+                // remove trailing trivia from places where we would want to insert the parameter list before a line break
+                originalDeclarationNode.Identifier.WithoutTrailingTrivia(),
+                originalDeclarationNode.TypeParameterList?.WithoutTrailingTrivia(),
+                SyntaxFactory.ParameterList(SyntaxFactory.SeparatedList(propertiesToAddAsParams)),
+                // TODO: inheritance
+                baseList: null,
+                originalDeclarationNode.ConstraintClauses,
+                SyntaxFactory.Token(SyntaxKind.OpenBraceToken),
+                SyntaxFactory.List(membersToKeep),
+                SyntaxFactory.Token(SyntaxKind.CloseBraceToken),
+                semicolonToken: default)
+                .WithLeadingTrivia(modifiedClassTrivia)
+                .WithAdditionalAnnotations(Formatter.Annotation);
+
+            var changedDocument = await document.ReplaceNodeAsync(originalDeclarationNode, changedTypeDeclaration, cancellationToken).ConfigureAwait(false);
+            return changedDocument;
+        }
+
+        private static bool ShouldConvertProperty(MemberDeclarationSyntax member, INamedTypeSymbol containingType)
+        {
+            if (member is not PropertyDeclarationSyntax property)
+            {
+                return false;
+            }
+
+            if (property.Initializer != null || property.ExpressionBody != null)
+            {
+                return false;
+            }
+
+            var propAccessibility = CSharpAccessibilityFacts.Instance.GetAccessibility(member);
+
+            // more restrictive than internal (protected, private, private protected, or unspecified (private by default))
+            if (propAccessibility < Accessibility.Internal)
+            {
+                return false;
+            }
+
+            // no accessors declared
+            if (property.AccessorList == null)
+            {
+                return false;
+            }
+
+            // for class records and readonly struct records, properties should be get; init; only
+            // for non-readonly structs, they should be get; set;
```

Comment:
needs explanation as to why.  also, we shoudl strongly consider still allowing the user to change `get/set` properties, just with a message saying it could break thing.   thsi can come later.

Reply:
My plan for this was:
- If there are no `get/init` properties, convert all (public) `get/set` properties to `get/init` (and move to primary constructor)
- If there are some `get/init` properties already, then `get/public set` properties are converted to primary constructor parameters, but still have the property definition to override explicitly. The initializer is set to the primary constructor parameter. This is because if they already know/use the `init` property word, then they likely have the other properties as `set` for a reason.
- example:
```
public class C
{
  public int P {get; init;}
  public bool B {get; set;}
}
```
would transform to:
```
public record C(int P, bool B)
{
  public bool B {get; set;} = B;
}
```
Unfortunately this may further mess up the property initializer/param default case, as we couldn't support `public int Foo {get; set;} = *Some initializer that already exists*`. But, seeing as we don't support this right now anyways, I think it's reasonable to have this, as it's how the [language suggests](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record#positional-syntax-for-property-definition) we override the default accessors of a record.

For the non-primary constructor case, we'd follow a similar idea: If we see no `init` properties, then we convert all public `get` properties to `init` ones, otherwise we wouldn't change them.
I'll probably add this in a follow up PR, since it requires some major refactoring of the logic (all properties need to be analyzed as a whole, not individually)

@CyrusNajmabadi @davidwengier Does this make sense/sound reasonable?